src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,207 @@
+                    }
+                    else
+                    {
+                        Log($"Path '{path}' did not exist.");
+                    }
+                }
+                else
+                {
+                    Log($"Path '{path}' had cached timestamp '{time}'.");
+                }
+
+                if (latestTime < time)
+                {
+                    latestTime = time;
+                }
+            }
+
+            return latestTime;
+        }
+
+        private async Task CheckReferencesAsync<TUnresolvedReference, TResolvedReference>(string name, IResolvableReferencesService<TUnresolvedReference, TResolvedReference> service, List<string> inputs)
+            where TUnresolvedReference : IProjectItem, TResolvedReference
+            where TResolvedReference : class, IReference
+        {
+            Log($"Checking reference type '{name}'.");
+            if (service != null)
+            {
+                foreach (var resolvedReference in await service.GetResolvedReferencesAsync())
+                {
+                    var fullPath = await resolvedReference.GetFullPathAsync();
+                    Log($"Adding input reference ${fullPath}");
+                    inputs.Add(fullPath);
+                }
+            }
+        }
+
+        public async Task<bool> IsUpToDateAsync(BuildAction buildAction, TextWriter logger, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (buildAction != BuildAction.Build)
+            {
+                return false;
+            }
+
+            using (var access = await _projectLockService.ReadLockAsync(cancellationToken))
+            {
+                var project = await access.GetProjectAsync(_configuredProject, cancellationToken);
+
+                Log($"Starting check for project '{project.FullPath}'.");
+
+                var timestampCache = _fileTimestampCache != null ? _fileTimestampCache.Value.TimestampCache : new Dictionary<string, DateTime>(StringComparer.OrdinalIgnoreCase);
+
+                if (_projectItemsSchema == null)
+                {
+                    Log($"Skipping because '{typeof(IProjectItemSchemaService).FullName}' component could not be found.");
+                    Report.IfNotPresent(_projectItemsSchema);
+                    return false;
+                }
+
+                ConfigurationGeneral general = await _projectProperties.GetConfigurationGeneralPropertiesAsync();
+
+                if ((bool)await general.DisableFastUpToDateCheck.GetValueAsync())
+                {
+                    Log($"Disabled because the 'DisableFastUpToDateCheckProperty' property is true.");
+                    return false;
+                }
+
+                var inputs = new List<string>();
+
+                // add the project file
+                if (!string.IsNullOrEmpty(_configuredProject.UnconfiguredProject.FullPath))
```

Comment:
Yeah targets should be enough, thinking about it `MSBuildAllProjects` is only needed for `<Target>` because they can't say all targets are my input.